### PR TITLE
Implement RedisStore

### LINKED ISSUE #63: Implement a redis store
Create a store that can write cache entries in redis.

Either create a `RedisStore` (similar to `MemoryStore` already provided) or just an example of how to provide a custom `store` option that will use redis.

### PR DESCRIPTION
**Pull request by mohamedgomran**
Fixes #63

### CODE REVIEW

**Comment by mohamedgomran** [comment]
@RasCarlito Please have a quick review.

- I have exposed RedisStore to be imported and called with redis client `new RedisStore(client)`.
- I wanna ask a question
  - How to ignore bundling RedisStore file in case of browser?

**Comment by RasCarlito** [comment]
Hey!

Thanks a lot for your contribution! This is awesome 😃 

For your question about not bundling `RedisStore` for browser build we need to have different webpack entry points for browser and node, for now both use `src/index.js`.

* We could move logic from `index.js` to another file (maybe named `api.js`)
* Then have `index.js` and `index.node.js` which both require `api.js` and expose the same methods except that `index.node.js` would add `RedisStore`.
* In `webpack.config.js` we would [need to make the entry](https://github.com/RasCarlito/axios-cache-adapter/blob/master/webpack.config.js#L58) dynamic depending on the [target being equal to 'node'](https://github.com/RasCarlito/axios-cache-adapter/blob/master/webpack.config.js#L42)

Other notes/questions:
* Should `RedisStore` have unit tests? For now unit tests are executed in a browser, should we add another test config to be executed using node?
* Could you add an example in the readme like the [localforage example](https://github.com/RasCarlito/axios-cache-adapter#use-localforage-as-cache-store) also showing how to pass a custom `HASH_KEY` to `RedisStore` constructor?

Cheers 🍻

**Comment by mohamedgomran** [comment]
Thanks you @RasCarlito 

Thanks for the guide.

- Yes we should have unit test, It would be the same as memory module test specs.
- Will add an example in the readme.

**Comment by mohamedgomran** [comment]
@RasCarlito I added unit tests and edited README.md file. :smile: 
I covered the new file I have added 100% but the codecov is telling me that it's not 100% like master, I pulled the master and run the test and it's not 100%, How codecov has the master 100% as covered even it's not? :disappointed:

**Comment by RasCarlito** [comment]
Hey @mohamedgomran! Great work! 

Sorry I was really not available lately.

Tried to dig a bit into codecov but couldn't find any explanation either. Let's move forward with this instead of spending time getting that .67% of coverage 😄 

I'll just review one last time then we can merge into master to make a release 🚀

File: README.md [hunk 1]
```diff
@@ -412,12 +441,22 @@ axios at the same time.
 }
 ```
 
-All the other parameters will be passed directly to the [`axios.create`](https://github.com/mzabriskie/axios#creating-an-instance) method.
```
Comment (RasCarlito) [inline README.md]: Please don't remove this part of the readme

File: README.md [hunk 2]
```diff
@@ -412,12 +441,22 @@ axios at the same time.
...
+
+```js
+    // Using redis client https://github.com/NodeRedis/node_redis
+    const redis = require("redis")
+    const client = redis.createClient()
+```
 
 #### Returns
 
-`setup()` returns an instance of `axios` pre-configured with the cache adapter.
```
Comment (RasCarlito) [inline README.md]: Same here, don't remove this part of the readme

File: README.md [hunk 3]
```diff
@@ -412,12 +441,22 @@ axios at the same time.
...
-All the other parameters will be passed directly to the [`axios.create`](https://github.com/mzabriskie/axios#creating-an-instance) method.
+### RedisStore(client, [, hashKey])
+
+RedisStore allow you to cache requests on server using [redis](https://redis.io/).
+Create a `RedisStore` instance. Takes `client` (`RedisClient`) and optional `hashKey` (name of hashSet to be used in redis).
+
+#### client
+
+```js
+    // Using redis client https://github.com/NodeRedis/node_redis
```
Comment (RasCarlito) [inline README.md]: Should we add `redis` client as a peer dependency?
Or maybe just add a comment about which version is supported in the readme?

**Comment by purplekrayons** [comment]
Really looking forward to this release! Thanks for your efforts @mohamedgomran. Should be fantastic when combined with the scale of the [Azure Redis service](https://azure.microsoft.com/en-ca/services/cache/). It'll render the size limit problems in `localStorage` moot and will allow me to cache much more of my application.

File: README.md (hunk 3 above)
Comment (mohamedgomran) [reply README.md]: I'm with adding comment with the version supported till now.

File: README.md (hunk 2 above)
Comment (mohamedgomran) [reply README.md]: Done

**Comment by mohamedgomran** [comment]
@RasCarlito Sorry for being late, thanks for the review. All discussion have been resolved. :100:

**Comment by mohamedgomran** [comment]
> Really looking forward to this release! Thanks for your efforts @mohamedgomran. Should be fantastic when combined with the scale of the [Azure Redis service](https://azure.microsoft.com/en-ca/services/cache/). It'll render the size limit problems in `localStorage` moot and will allow me to cache much more of my application. Notably looking forward to layer of reliability it'll offer when choosing to read the cache if network errors occur when setting the `readOnError` config.

Thank you for your words :)

It also would resolve the problem of multiple instance applications (Like using docker-compose to scale up your service using multiple replicated containers) each replica of application has a separate memory and memoryStore would not be shared across them.

**Comment by RasCarlito** [comment]
Hey! I'm on vacation so I haven't had time checking in on this 😅 
Thanks for the final modifications. This looks good.

I'll merge it and release it quickly 🚀

**Comment by mohamedgomran** [comment]
@RasCarlito Oh, I was about to contact you, welcome back. Thanks for the merge :1st_place_medal:

**Pull request merged**